### PR TITLE
Feature/initial feature

### PR DESCRIPTION
**Pull request by DiegoCabreraN**
#### What does this PR do?
- this PR Initializes the Progressions API Repository
- Connects to Twilio
- Obtains Data from Zoho
- Obtains Data from PostgreSQL
- Sends a Whatsapp to the relevant account
#### Where should the reviewer start?
:checkered_flag: /index.js
#### How should this be manually tested?
1.  Install all NPM dependencies
```npm install```
2. Add the environment varibles
3. Start the server
```npm start```
4. Start ngrok
```ngrok http 4000```
5. Copy ngrok URL and paste it inside the Twilio's Console inside the Whatsapp Sandbox Configuration fields.
6. Join the sandbox by sending the message 
```join massage-respect``` from your phone to the number: +1 415 523 8886
7. Send a POST request to the ngrok URL ex:
```POST xxxxxxxx.ngrok.io/send```
8. If you are in a progression day you can send a word and follow the flux, else you have to wait until a progression date or to update the row inside postgreSQL
#### Any background context you want to provide?
- A progression is a 2-week report for clients
- It is done by sending a WhatsApp to the client with the info he's related to
- In other words this is an automatization for the progression workflow.
- The only restriction we have is the approval for our [Twilio WhatsApp number from Facebook](https://www.twilio.com/docs/sms/whatsapp/api#facebook-business-manager-account)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// # Imports # //
+
+import express from 'express';
```
Comment (walbertoibarra) [inline index.js]: Using Babel just for being able to use ES6 modules is an overkill, if you really like the feature I'd suggest to use it with Node.js directly, e.g. https://stackoverflow.com/a/45854500/4301141.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import bodyParser from 'body-parser';
+import {sendToWhatsapp} from './src/services/whatsapp';
+import {getDataFromPostgreSQL} from './src/services/bodies';
```
Comment (walbertoibarra) [inline index.js]: Using Babel just for being able to use ES6 modules is an overkill, if you really like the feature I'd suggest to use it with Node.js directly, e.g. https://stackoverflow.com/a/45854500/4301141 or just use CommonJS import/export style.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// --- Create the Server --- //
+const app = express();
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(bodyParser.json());
+
+// Set Port //
+const PORT = 4000;
+
+// Define Post Method //
+app.post('/subscriptions', (req, res) => {
```
Comment (walbertoibarra) [inline index.js]: Please move route definitions into the `src/routes` directory, leave only server setup on this file.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    "clear": "^0.1.0",
+    "dotenv": "^8.2.0",
+    "express": "^4.17.1",
+    "handlebars": "^4.7.3",
+    "mongodb": "^3.5.2",
+    "mongoose": "^5.8.10",
+    "ngrok": "^3.2.7",
+    "nodemon": "^2.0.2",
+    "pg": "^7.18.1",
+    "twilio": "^3.39.3"
```
Comment (walbertoibarra) [inline package.json]: Please freeze dependencies as it helps avoid awkward errors on future deployments.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "node_rest",
+  "version": "1.0.0",
+  "main": "index.js",
+  "scripts": {
+    "start": "nodemon ./index.js --exec babel-node -e js"
```
Comment (walbertoibarra) [inline package.json]: Could you move this script to `dev` and make `start` run with `node`?

File: src/Services/bodies.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  password: env.dbPassword,
+  port: env.dbPort,
+});
+function diff_weeks(t2, t1) { 
+  var diff =(t2.getTime() - t1.getTime()) / (1000 * 3600 * 24);
+  diff = Math.abs(Math.round(diff)) 
+  if(diff%14 === 0){
+    return diff/14;
+  }
+  return undefined;
```
Comment (walbertoibarra) [inline src/Services/bodies.js]: No need to actually return `undefined` here.

File: src/Services/bodies.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      else if(arg === 'answeredGood'){
+        data[`${arg}`] = '98%';
+      }
+      else if(arg === 'alias'){
+        data[`${arg}`] = rowData.alias;
+      }
+    }
+  }
+  const template = handlebars.compile(fileString)
+  const newStr = template(data);
```
Comment (walbertoibarra) [inline src/Services/bodies.js]: Unnecessary constant declaration.

File: src/Services/bodies.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+const fs = require('fs');
+const PoolSQL  = require('pg').Pool; 
+const pool = new PoolSQL({
+  user: env.dbUser,
+  host: env.dbHost,
+  database: env.dbName,
+  password: env.dbPassword,
+  port: env.dbPort,
+});
+function diff_weeks(t2, t1) {
```
Comment (walbertoibarra) [inline src/Services/bodies.js]: We don't do that here, please stick to naming your variables/functions with camel case.

File: index.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import {getDataFromPostgreSQL} from './src/services/bodies';
+
+
+// --- Create the Server --- //
+const app = express();
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(bodyParser.json());
+
+// Set Port //
+const PORT = 4000;
```
Comment (walbertoibarra) [inline index.js]: Move to `src/config`, read from `process.env` and default it to `4000` there.

File: src/Services/updatePostgreSQL.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    });
+    data = data.concat(subresp.data.data); 
+    more_pages = subresp.data.info.more_records;
+    page += 1; 
+  } 
+  data = data.filter(item => item.Account !== undefined && item.Account !== null);
+  data = data.filter(item => item.Account.id !== undefined && item.Account.id !== null);
+  data = data.filter(item => item.id !== undefined && item.id !== null);
+  data = data.filter(item => item.Alias !== undefined && item.Alias !== null && !item.Alias.includes('Prueba') && !item.Alias.includes('Demo'));
+  data = data.filter(item => item.Setup_Date !== undefined && item.Setup_Date !== null);
```
Comment (walbertoibarra) [inline src/Services/updatePostgreSQL.js]: Most of these filter functions are almost identical, please refactor to use the same function, also, change how you check that the key has value.

File: src/Services/bodies.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  const rows = await pool.query(`SELECT * FROM clients WHERE "clientspersonalname" IS NOT NULL`);
+  for(var i = 0; i<rows.rows.length; i++){
+    const day = new Date().getDate();
+    const month = new Date().getMonth();
+    const year = new Date().getFullYear();
+    const today = new Date(year,month,day,0,0,0,0);
+    const instalationDate = new Date(rows.rows[i].instalationdate);
+    const week = diff_weeks(today,instalationDate);
+    if(week){
+      const tmpFile = fs.readFileSync(`src/templates/0.txt`, (err, data) => {
```
Comment (walbertoibarra) [inline src/Services/bodies.js]: IO operations are one of the most expensive to have, and making it synchronous is terrible since it would block our whole Node.js process, please consider caching those files in memory, and do change those calls to their asynchronous equivalent.

File: src/Services/bodies.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+export async function sendData(req){ 
+  const rows = await pool.query(`SELECT * FROM clients WHERE "phone" = '${req.split(':')[1].replace('+','')}'`);
+  const day = new Date().getDate();
+  const month = new Date().getMonth();
+  const year = new Date().getFullYear();
+  const today = new Date(year,month,day,0,0,0,0);
+  const instalationDate = new Date(rows.rows[0].instalationdate);
+  const week = diff_weeks(today,instalationDate);
+  if(week){
+    const tmpFile = fs.readFileSync(`src/templates/${week}.txt`, (err, data) => {
```
Comment (walbertoibarra) [inline src/Services/bodies.js]: Change this call to its asynchronous equivalent.

File: src/Services/whatsapp.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        to: req.body.From
+      })
+      .catch(error=>{
+        console.log(error);
+      });
+  }
+}
+
+function showError(){
+  const tmpFile = fs.readFileSync(`src/templates/error.txt`, (err, data) => {
```
Comment (walbertoibarra) [inline src/Services/whatsapp.js]: Change this call to its asynchronous equivalent.

File: src/config/index.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
+require('dotenv').config();
```
Comment (walbertoibarra) [inline src/config/index.js]: This should be the first line to run in the process since some libraries change behavior on some env vars values, move it to `index.js`.

File: src/routes/crmRoutes.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import {addNewContact,
```
Comment (walbertoibarra) [inline src/routes/crmRoutes.js]: Is this file being imported anywhere in the project?

File: index.js (hunk 2 above)
Comment (Angel-Mu) [reply index.js]: A.k.a use `require('module')` `module.exports` from `CommonJS`.  https://flaviocopes.com/commonjs/

File: package.json (hunk 5 above)
Comment (Angel-Mu) [reply package.json]: Agree with this... `start` should run only with node or process managers if case, but nodemon is a daemon that listens to changes in the code, useful while you're coding

File: package.json (hunk 4 above)
Comment (DiegoCabreraN) [reply package.json]: hahaha y como se hace eso?

File: package.json (hunk 5 above)
Comment (DiegoCabreraN) [reply package.json]: - [X] Done

File: src/Services/bodies.js (hunk 6 above)
Comment (DiegoCabreraN) [reply src/Services/bodies.js]: - [X] Fixed

File: src/bodies/bodyGenerator.js [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+  },
+  {
+    name: 'customerAge',
+    filepath: './src/templates/reports/queries/customerAge.txt'
+  }
+]
+function diffWeeks(t2, t1) { 
+  var diff =(t2.getTime() - t1.getTime()) / (1000 * 3600 * 24);
+  diff = Math.abs(Math.round(diff)) 
+  if(diff%14 === 0){
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Shouldn't this return something? this if will cause an undefined return which is pretty ugly to have a function that doesn't return consistently, ie. this iff is returning a number if this condition passes, but... if not, nothing get's resolved. Also, according with the name of the function I get that this should return the diff of weeks, and based on parameters, I get, weeks of difference between two dates. A undefined return is not what is meant to be.

File: src/bodies/bodyGenerator.js [hunk 17]
```diff
@@ -0,0 +1,240 @@
+const env = require('../config/index');
+const handlebars = require('handlebars');
+const fs = require('fs');
+const util = require('util');
+const readFile = util.promisify(fs.readFile);
+const poolSQL  = require('pg').Pool;
+const templateMapper = [
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: create a `constants || mappers` folder and move this dictionaries into it. This way you will only require them when you need them and have the code a bit more organized

File: src/bodies/bodyGenerator.js [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+        const whatsappMessage = await mountData(tmpFile, today, rows[i], pool);
+        bodies.push({
+          alias : rows[i].alias,
+          phone : rows[i].phone,
+          body : whatsappMessage
+        });
+      }catch(e){}
+    } 
+  }
+  console.log(bodies)
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Remove this console log, or print something more useful instead. This will only cause a mess in logs and very hard to debug.

File: src/bodies/bodyGenerator.js [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+  if(diff%14 === 0){
+    return diff/14;
+  }
+}
+function calcStartDate(date){
+  resp = new Date(date);
+  resp.setDate( date.getDate() - 14); 
+  return resp;
+}
+async function mountQuery(name, args, pool){
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: name this function differently, doesn't reflect what it does!

File: src/bodies/bodyGenerator.js [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+  const queryData = queryMapper.find(item => item.name === name);
+  const tmpQueryFile = await readFile(queryData.filepath, 'utf-8');
+  const template = handlebars.compile(tmpQueryFile);
+  compiledQuery = template(args);
+  queryResult = await pool.query(compiledQuery);
+  return queryResult.rows;
+}
+async function mountData(fileString, today, rowData, pool){
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Same here, is not clear what `mountData` means or what this does

File: src/bodies/bodyGenerator.js [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+  //let startDate = calcStartDate(today);
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Remove commented code. I know is for testing purposes, but that should be only in your environment, not the code we want to promote.

File: src/bodies/bodyGenerator.js [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+    filepath: './src/templates/reports/biweekly/progresion3.txt',
+    order: 3
+  },
+  {
+    description: 'Fourth Progression Template',
+    filepath: './src/templates/reports/biweekly/progresion4.txt',
+    order: 4
+  }
+]
+const queryMapper = [
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Same for this mapper.

File: src/bodies/bodyGenerator.js [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+    return diff/14;
+  }
+}
+function calcStartDate(date){
+  resp = new Date(date);
+  resp.setDate( date.getDate() - 14); 
+  return resp;
+}
+async function mountQuery(name, args, pool){
+  const queryData = queryMapper.find(item => item.name === name);
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: this doesn't return a queryData, if finds its value only. It's very confusing, I thought at first that this was resolving the query but it doesn't. Please use `queryDefinition` or just `query`.

File: src/bodies/bodyGenerator.js [hunk 24]
```diff
@@ -0,0 +1,240 @@
+const env = require('../config/index');
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: replace this to `const config = require('../config')`; as it doesn't only resolve the env vars, you can export also different kind of configurations not coming directly from env. 
Also a better practice is just to deconstruct and get what you really need so you don't waste too much memory obtaining the whole object every time. ie.
```javascript
const { myProp } = require('../config');
```

File: src/bodies/bodyGenerator.js [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+const handlebars = require('handlebars');
+const fs = require('fs');
+const util = require('util');
+const readFile = util.promisify(fs.readFile);
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: `fs` already comes with sync if that's what you want. No need to promisify this. Is an overhead for something already comes with nodejs
https://nodejs.org/api/fs.html#fs_fs_readfilesync_path_options

File: src/bodies/bodyGenerator.js [hunk 26]
```diff
@@ -0,0 +1,240 @@
...
+      const queryArgs = {
+        startDate: startDate,
+        endDate: endDate,
+        lpid: rowData.lpid,
+        installDate: installDate
+      };
+      switch (arg){
+
+        case 'subscriptionName':
+          data[`${arg}`] = rowData.alias;
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Unnecessary use of string template here. It could be only as `data[arg] = ` Please change this for all of them.

File: src/bodies/bodyGenerator.js [hunk 27]
```diff
@@ -0,0 +1,240 @@
...
+          break;
+
+        case 'name':
+          data[`${arg}`] = rowData.clientspersonalname;
+          break;
+
+        case 'totalCustomers':
+          queryResult = await mountQuery(arg, queryArgs, pool); 
+          data[`${arg}`] = parseInt(queryResult[0].count);
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: So unsafe to use fixed positions in an array, are you sure that the result you need always comes at the `0` index? where's this validation. Also, is better not to use fixed positions. It may get you an undefined.
Two ways to go from here, either you ensure your query resolved every time one single record and return only 1 row but as an object instead, or deconstruct the array and validate it exists. ie. 
```javascript
const [queryResult] = await mountQuery(arg, queryArgs, pool);
data[arg] = parseInt(queryResult ? queryResult.count : 0);
```

Another thing I noticed, `queryResult` is declared outside this switch and never gets used out of this context. Better to declared as const every time instead. So this way you avoid to have it outside of this scope.

File: src/bodies/bodyGenerator.js [hunk 28]
```diff
@@ -0,0 +1,240 @@
...
+        case 'newCustomersBiweekly': 
+          queryResult = await mountQuery(arg ,queryArgs, pool);
+          clientsAsked = 0;
+          queryResult.forEach(item => {
+            clientsAsked += parseInt(item.count);
+          })
+          isFirstTime = Math.round((parseInt(queryResult[0].count)/clientsAsked)*100);
+          isNotFirstTime = 100-isFirstTime;
+          data[`${arg}`] = clientsAsked;
+          data['isFirstTime'] = isFirstTime;
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Just FYI, you can also use dot notation here.  `data.isFirstTime = isFirstTime`.
And also you can construct using spread operator (probably for another day).

File: src/bodies/bodyGenerator.js [hunk 29]
```diff
@@ -0,0 +1,240 @@
...
+          data['isNotFirstTime'] = isNotFirstTime; 
+          break;
+
+        case 'customerPerception':
+          queryResult = await mountQuery(arg ,queryArgs, pool);
+          clientsAsked = 0;
+          let thinkGoodService, thinkBadService = 0;
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Too much code for a case, consider to create a function and call it here. That function could return and object and you only used it in here.

File: src/bodies/bodyGenerator.js [hunk 30]
```diff
@@ -0,0 +1,240 @@
...
+                thinkBadService += parseInt(item.count);
+                break;
+            }
+          })
+          data[`${arg}`] = clientsAsked;
+          data['thinkGoodService'] = Math.round((thinkGoodService/clientsAsked)*100);
+          data['thinkBadService'] = Math.round((thinkBadService/clientsAsked)*100);
+        
+        case 'friendRecommendation': 
+          queryResult = await mountQuery(arg ,queryArgs, pool);
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Same here. Too much code for a case block

File: src/bodies/bodyGenerator.js [hunk 31]
```diff
@@ -0,0 +1,240 @@
...
+    if (item){
+      tmpFile = await readFile(`${item.filepath}`, 'utf-8'); 
+      try{
+        const whatsappMessage = await mountData(tmpFile, today, rows[i], pool);
+        bodies.push({
+          alias : rows[i].alias,
+          phone : rows[i].phone,
+          body : whatsappMessage
+        });
+      }catch(e){}
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: This won't tell us anything, it will just fail and no one will notice. If the purpose is only preventing the failure, I will prefer to fail and to know what caused the failure.

File: src/bodies/bodyGenerator.js [hunk 32]
```diff
@@ -0,0 +1,240 @@
...
+  const bodies = []; 
+  for(var i = 0; i<rows.length; i++){
+    console.log(rows[i].phone)
+    //Test lpid
+    rows[i].lpid = 564;
+    // 
+    const instalationDate = new Date(rows[i].instalationdate);
+    const today = new Date(new Date().getFullYear(),new Date().getMonth(),new Date().getDate());
+    const week = diffWeeks(today,instalationDate);
+    var tmpFile = undefined;
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: WHAT??? Why?? Avoid using vars, never assign and undefined, I see this only used once inside the if clause, so just declare it in there.

File: src/bodies/bodyGenerator.js [hunk 33]
```diff
@@ -0,0 +1,240 @@
...
+    database: env.dbName,
+    password: env.dbPassword,
+    port: env.dbPort,
+  });
+  rows = await mountQuery('getClients',{phone:env.testPhone}, pool);
+  const bodies = []; 
+  for(var i = 0; i<rows.length; i++){
+    console.log(rows[i].phone)
+    //Test lpid
+    rows[i].lpid = 564;
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: Hardcoded value, Same here, I know is for testing purposes, but consider better adding a env var which contains this kind of values, if exists, assign it, otherwise continue with the normal execution.

File: src/bodies/bodyGenerator.js (hunk 25 above)
Comment (DiegoCabreraN) [reply src/bodies/bodyGenerator.js]: Actually I was using the sync version, but Waldo asked me to use the async equivalent

File: src/bodies/bodyGenerator.js (hunk 16 above)
Comment (angel-mu-xb) [reply src/bodies/bodyGenerator.js]: better to `return (diff%14 === 0) ? diff/14 : 0` save lines of code

File: src/bodies/bodyGenerator.js [hunk 34]
```diff
@@ -0,0 +1,196 @@
+const handlebars = require('handlebars');
+const fs = require('fs');
+const poolSQL  = require('pg').Pool;
+const { templateMapper, queryMapper} = require('../constants/mappers');
+const { dbUser, dbHost, dbName, dbPassword, dbPort, testPhone, testLpid } = require('../config');
+const readFile = (filename) => {
+  return new Promise((resolve, reject) => {
+      fs.readFile(filename, 'utf-8', (err, data) =>{
+          if (err)
```
Comment (Angel-Mu) [inline src/bodies/bodyGenerator.js]: please be consistent with your indentation here.
